zenoh-flow/src/traits.rs

```diff
@@ -22,49 +22,194 @@ use std::any::Any;
 pub trait Node {
+    /// This method is used to initialize the state of the node.
+    /// It is called by the Zenoh Flow runtime when initializing the data flow
+    /// graph.
+    /// An example of node state is files that should be opened, connection
+    /// to devices or internal configuration.
+    ///
+    /// # Errors
+    /// If it fails to initialize an error variant will be returned.
     fn initialize(&self, configuration: &Option<Configuration>) -> ZFResult<State>;
 
+    /// This method is used to finalize the state of the node.
+    /// It is called by the Zenoh Flow runtime when tearing down the data
+    /// flow graph.
+    ///
+    /// An example of node state to finalize is files to be closed,
+    /// clean-up of libraries, or devices.
+    ///
+    /// # Errors
+    /// If it fails to finalize an error variant will be returned.
     fn finalize(&self, state: &mut State) -> ZFResult<()>;
 }
 
+/// The `Operator` trait represent an Operator inside Zenoh Flow.
 pub trait Operator: Node + Send + Sync {
+    /// This method is called when data is received on one or more inputs.
+    /// The result of this method is use as discriminant to trigger the
+    /// operator's run function.
+    /// The operator can access to its context and its state during execution.
+    ///
+    /// The received data is provided as [`InputToken`](`InputToken`) that
+    /// represent the state of the associated port.
+    /// Based on the tokens and on the data users can decide if trigger
+    /// the run or not.
+    /// The commodity function [`default_input_rule`](`default_input_rule`)
+    /// can be used if the operator should be triggered based on KPN rules.
+    ///
+    /// # Errors
+    /// If something goes wrong during execution an error
+    /// variant will be returned.
     fn input_rule(
         &self,
         context: &mut Context,
         state: &mut State,
         tokens: &mut HashMap<PortId, InputToken>,
     ) -> ZFResult<bool>;
 
+    /// This method is the actual one processing the data.
+    /// It is triggered based on the result of the `input_rule`.
+    /// As operators are computing over data,
+    /// *I/O should not be done in the run*.
+    ///
+    /// The operator can access to its context and its state during execution.
+    /// The result of a computation can also not provide any output.
+    /// When it does provide output the `PortId` used should match the one
+    /// defined in the descriptor for the operator. Any not matching `PortId`
+    /// will be dropped.
+    ///
+    /// # Errors
+    /// If something goes wrong during execution an error
+    /// variant will be returned.
     fn run(
         &self,
         context: &mut Context,
         state: &mut State,
         inputs: &mut HashMap<PortId, DataMessage>,
     ) -> ZFResult<HashMap<PortId, Data>>;
 
+    /// This method is called after the run, and can be used for
+    /// further analysis and adjustment over the computed data.
```

Comment:
This method is called after the run, its main purpose is to check for missed _local_ deadlines. It can also be used for further analysis and…